src/blockchain/Blockchain.ts

```diff
@@ -141,28 +211,130 @@ export default class Blockchain implements IBlockchainApi {
+  }
+
+  /**
+   * [ASYNC] Signs and submits the SubmittableExtrinsic with optional resolution and rejection criteria.
+   *
+   * @param identity The [[Identity]] that we sign and potentially re-sign the tx with.
+   * @param tx The generated unsigned [[SubmittableExtrinsic]] to submit.
+   * @param opts Partial optional criteria for resolving/rejecting the promise.
+   * @returns Promise result of The Extrinsic.
+   *
+   */
   public async submitTx(
     identity: Identity,
     tx: SubmittableExtrinsic,
-    opts: SubscriptionPromiseOptions = {}
+    opts?: Partial<SubscriptionPromiseOptions>
   ): Promise<SubmittableResult> {
     const signedTx = await this.signTx(identity, tx)
-    return Blockchain.submitSignedTx(signedTx, opts)
+    return this.submitSignedTx(identity, signedTx, opts)
+  }
+
+  /**
+   * [ASYNC] Retrieves the Nonce for Transaction signing for the specified account and increments the in accountNonces mapped Index.
+   *
+   * @param accountAddress The address of the identity that we retrieve the nonce for.
+   * @returns [[BN]] representation of the Tx nonce for the identity.
+   *
+   */
+  public async getNonce(accountAddress: string): Promise<BN> {
+    let nonce = this.accountNonces.get(accountAddress)
+    if (!nonce) {
+      const chainNonce = await this.api.rpc.system
+        .accountNextIndex(accountAddress)
+        .catch((reason) => {
+          log.error(
+            `On-chain nonce retrieval failed for account ${accountAddress}\nwith reason: ${reason}`
+          )
+          throw Error(`Chain failed to retrieve nonce for : ${accountAddress}`)
+        })
+      const secondQuery = this.accountNonces.get(accountAddress)
+      nonce = BN.max(chainNonce, secondQuery || new BN(0))
+    }
+    this.accountNonces.set(accountAddress, nonce.addn(1))
+    return nonce
   }
 
-  public async getNonce(accountAddress: string): Promise<Index> {
-    return this.api.rpc.system.accountNextIndex(accountAddress)
+  /**
+   * [ASYNC] Re-signs the given [[SubmittableExtrinsic]] with an updated Nonce.
+   *
+   * @param identity The [[Identity]] to re-sign the Tx with.
+   * @param tx The previously with recoverable Error failed Tx.
+   * @returns Original Tx, injected with signature payload with updated nonce.
+   *
+   */
+  public async reSignTx(
+    identity: Identity,
+    tx: SubmittableExtrinsic
+  ): Promise<SubmittableExtrinsic> {
+    this.accountNonces.delete(identity.address)
+    const nonce: BN = await this.getNonce(identity.address)
+    const signerPayload: SignerPayloadJSON = this.api
+      .createType('SignerPayload', {
+        method: tx.method.toHex(),
+        nonce,
+        genesisHash: this.api.genesisHash,
+        blockHash: this.api.genesisHash,
+        runtimeVersion: this.api.runtimeVersion,
+        version: this.api.extrinsicVersion,
+      })
+      .toPayload()
+    tx.addSignature(
+      identity.address,
+      this.api
```

Comment:
daaaaamn I thought you could just do `tx.sign()` again 😣

Reply:
Only addSignature doesn't check whether the extrinsic was already signed, and just injects the data, regardless of existence of the signature.

Reply:
that sucks man, so much code just to get around this